tests/pln/TemporalReasoningUTest.cxxtest

```diff
@@ -0,0 +1,171 @@
+			cur_dir = cur_ppp_dir + "/tests/pln/rules",
+			src_rule_dir = cur_ppp_dir + "/opencog/pln/rules";
+		std::vector<std::string> load_paths = {cur_ppp_dir, cur_dir, src_rule_dir};
+		for (std::string& p : load_paths)
+		{
+			std::string eval_str = std::string("(add-to-load-path \"") + p + std::string("\")");
+			eval.eval(eval_str);
+		}
+
+		eval.eval("(use-modules (opencog))");
+		eval.eval("(use-modules (opencog spacetime))");
+		eval.eval("(use-modules (opencog ure))");
+		eval.eval("(use-modules (opencog pln))");
+	}
+
+	~TemporalReasoningUTest()
+	{
+		// Erase the log file if no assertions failed.
+		if (!CxxTest::TestTracker::tracker().suiteFailed())
+			std::remove(logger().get_filename().c_str());
+	}
+
+	void setUp();
+	void tearDown();
+	void load_scm_files(std::vector<std::string>);
+	void test_cogscm_temporal_deduction();
+};
+
+void TemporalReasoningUTest::tearDown()
+{
+	as.clear();
+}
+
+void TemporalReasoningUTest::setUp()
+{
+}
+
+void TemporalReasoningUTest::load_scm_files(std::vector<std::string> files)
+{
+	std::string hd("(load-from-path \"");
+	std::string tl("\")");
+	for (auto f: files) {
+		std::string result = eval.eval(hd + f + tl);
+		logger().debug() << "[TemporalReasoningUTest] Loaded \"" << f << "\":"
+		                 << std::endl << result;
+	}
+}
+
+/**
+ * Test cogscm temporal-deduction-rule defined in:
+ * opencog/pln/rules/temporal/back-predictive-implication-scope-deduction-cogscm.scm
+ */
+void TemporalReasoningUTest::test_cogscm_temporal_deduction()
+{
+	logger().debug("BEGIN TEST: %s", __FUNCTION__);
+
+	// load KB
+	load_scm_files({"tests/pln/rules/temporal-deduction.scm"});
+
+	// Load rule
+	std::string rs = eval.eval("(pln-load-rule 'back-predictive-implication-scope-deduction-cogscm)");
+
+	Handle target = eval.eval_h("target");
+	Handle results = eval.eval_h("(pln-bc target)");
+
+	Handle expected = eval.eval_h
+		("(SetLink"
+		 "(BackPredictiveImplicationScopeLink"
+		 "	(VariableNode \"$X\")"
+		 "	(SLink"
+		 "		(SLink"
+		 "			(ZLink)))"
+		 "	(BackSequentialAndLink"
+		 "		(SLink"
+		 "			(ZLink))"
+		 "		(EvaluationLink"
+		 "			(PredicateNode \"P\")"
+		 "			(VariableNode \"$X\"))"
+		 "		(ExecutionLink"
+         "			(SchemaNode \"A2\")))"
+    	 "	(EvaluationLink"
+      	 "		(PredicateNode \"R\")"
+      	 "		(VariableNode \"$X\")))"
+		 "(BackPredictiveImplicationScopeLink"
+		 "	(VariableNode \"$X\")"
+		 "	(SLink" 
+		 "		(SLink"
+		 "			(ZLink)))"
+		 "	(BackSequentialAndLink"
+		 "		(SLink"
+		 "			(ZLink))"
+		 "  	(AndLink"
+		 "			(EvaluationLink"
+		 "				(PredicateNode \"C\")"
+		 "				(VariableNode \"$X\"))"
+		 "			(ExecutionLink"
+		 "				(SchemaNode \"A1\")))"
+		 "		(ExecutionLink"
+         "			(SchemaNode \"A2\")))"
+    	 "	(EvaluationLink"
+      	 "		(PredicateNode \"R\")"
+      	 "		(VariableNode \"$X\"))))");
+		   
+	std::cout << "results = " << results->to_string();
+	std::cout << "expected = " << expected->to_string();
+
+	HandleSet sss;
```

Comment:
`sss` is no longer needed.

Reply:
It is because `(opencog spacetime)` is not installed, I will try to add that to the config file

Reply:
Oh, of course.  Then you have the option of making spacetime optional or not.  If spacetime is optional, then you should configure CMake to disable temporal reasoning unit tests.  If spacetime is not optional then the CMake config should require it.  I've been pondering that for a while already and I don't know which way is best.  It seems the simplest for us would be to make spacetime required, but that adds a slight annoyance to the user who may not want to use temporal reasoning.  What do you think we should do?

Reply:
Since it is just us using the temporal reasoning, I can disable it for now and we can enable it once we used it on rocca I think.

Reply:
Maybe it's better enforce the spacetime dependency for the time being though.  Not having spacetime installed while invoking `pln-load-rule` on a temporal rule is gonna crash PLN.  If we make temporal reasoning optional in PLN we should make it in a way that it doesn't crash.  This is possible but a bit of a pain (I don't know how to do it as of right now).  So basically what I'm saying is that making spacetime mandatory is safer to the user, and safety is more important than inconvenience.